src/decode_url.rs

```diff
@@ -0,0 +1,62 @@
+use crate::encode_url::encode_url;
+use napi::bindgen_prelude::*;
+use std::borrow::Cow;
+
+#[inline]
+fn from_hex_digit(digit: u8) -> Option<u8> {
+  match digit {
+    b'0'..=b'9' => Some(digit - b'0'),
+    b'A'..=b'F' => Some(digit - b'A' + 10),
+    b'a'..=b'f' => Some(digit - b'a' + 10),
+    _ => None,
+  }
+}
+
+pub fn decode_url_impl(url: &str) -> Cow<[u8]> {
+  let mut url_bytes = url.as_bytes();
+  let mut result = Vec::with_capacity(url_bytes.len());
```

Comment:
I think we can do a preliminary check before this line to see if the URL contains `%`. Otherwise, there is no need to allocate memory.